be/src/service/staros_worker.cpp

```diff
@@ -7,87 +7,199 @@
 #include "common/config.h"
 #include "common/logging.h"
 #include "fmt/format.h"
 
 namespace starrocks {
 
+namespace fslib = staros::starlet::fslib;
+
 std::ostream& operator<<(std::ostream& os, const staros::starlet::ShardInfo& shard) {
     return os << "Shard{.id=" << shard.id << " .uri=" << shard.obj_store_info.s3_obj_store.uri << "}";
 }
 
 absl::Status StarOSWorker::add_shard(const ShardInfo& shard) {
+    std::unique_lock l(_mtx);
     LOG(INFO) << "Adding " << shard;
-    std::lock_guard l(_mtx);
-    _shards[shard.id] = shard;
+    auto ret = _shards.try_emplace(shard.id, ShardInfoDetails(shard));
+    if (!ret.second) {
+        return absl::AlreadyExistsError(fmt::format("Shard id '{}' already exists!", shard.id));
+    }
     return absl::OkStatus();
 }
 
 absl::Status StarOSWorker::remove_shard(const ShardId id) {
+    std::unique_lock l(_mtx);
     LOG(INFO) << "Removing " << id;
-    std::lock_guard l(_mtx);
     _shards.erase(id);
     return absl::OkStatus();
 }
 
-StatusOr<staros::starlet::ShardInfo> StarOSWorker::get_shard_info(ShardId id) {
-    std::lock_guard l(_mtx);
+absl::StatusOr<staros::starlet::ShardInfo> StarOSWorker::get_shard_info(ShardId id) const {
+    std::shared_lock l(_mtx);
     auto it = _shards.find(id);
     if (it == _shards.end()) {
-        return Status::NotFound(fmt::format("failed to get shardinfo {}", id));
+        return absl::NotFoundError(fmt::format("failed to get shardinfo {}", id));
     }
-    return it->second;
+    return it->second.shard_info;
 }
 
-std::vector<staros::starlet::ShardInfo> StarOSWorker::shards() {
-    std::lock_guard l(_mtx);
+std::vector<staros::starlet::ShardInfo> StarOSWorker::shards() const {
     std::vector<staros::starlet::ShardInfo> vec;
+    vec.reserve(_shards.size());
+
+    std::shared_lock l(_mtx);
     for (const auto& shard : _shards) {
-        vec.emplace_back(shard.second);
+        vec.emplace_back(shard.second.shard_info);
     }
     return vec;
 }
 
-absl::StatusOr<staros::starlet::WorkerInfo> StarOSWorker::worker_info() {
+absl::StatusOr<staros::starlet::WorkerInfo> StarOSWorker::worker_info() const {
     staros::starlet::WorkerInfo worker_info;
-    std::lock_guard l(_mtx);
+
+    std::shared_lock l(_mtx);
     worker_info.worker_id = _worker_id;
     worker_info.service_id = _service_id;
     worker_info.properties["port"] = std::to_string(config::starlet_port);
-    for (auto&& [k, _] : _shards) {
-        worker_info.shards.insert(k);
+    for (auto& iter : _shards) {
+        worker_info.shards.insert(iter.first);
     }
     return worker_info;
 }
 
 absl::Status StarOSWorker::update_worker_info(const staros::starlet::WorkerInfo& new_worker_info) {
-    std::lock_guard l(_mtx);
+    std::unique_lock l(_mtx);
     _service_id = new_worker_info.service_id;
     _worker_id = new_worker_info.worker_id;
     return absl::OkStatus();
 }
 
+absl::StatusOr<std::shared_ptr<fslib::FileSystem>> StarOSWorker::get_shard_filesystem(ShardId id,
```

Comment:
maybe we can move this logic into starlet, so we can make be very easy?

Reply:
It can, but I would like it to be moved until we are working on the separating deployment mode. Right now it makes little different, and starlet types.h can be as simple as just data passing through.